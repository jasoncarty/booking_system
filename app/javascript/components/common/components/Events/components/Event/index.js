import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';

import EventListAttendees from './../EventListAttendees';
import EventActions from './../EventActions';

class Event extends Component {
  static propTypes = {
    id: PropTypes.number.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string,
    startsAt: PropTypes.string.isRequired,
    eventAttendees: PropTypes.arrayOf(PropTypes.shape({})),
    attendances: PropTypes.arrayOf(PropTypes.number),
    onDeleteSuccess: PropTypes.func.isRequired,
    onDeleteError: PropTypes.func.isRequired,
    adminSection: PropTypes.bool
  }

  state = {
    eventAttendees: [],
    attendances: []
  }

  componentDidMount() {
    this.setState({
      eventAttendees: this.props.eventAttendees,
      attendances: this.props.attendances
    });
  }

  getReserves = (eventAttendees) =>
    eventAttendees.filter((attendee) => attendee.reserve);

  getAttendees = (eventAttendees) =>
    eventAttendees.filter((attendee) => !attendee.reserve);

  handleUpdateAttendance = ({ data: { eventAttendees, attendances } }) =>
    this.setState({ eventAttendees, attendances })

  handleCancelError = (res) =>
    console.log('---handleCancelError----', res)

  handleJoinError = (res) =>
    console.log('---handleJoinError----', res)

  render() {
    const {
      id,
      name,
      description,
      startsAt,
      adminSection,
      onDeleteSuccess,
      onDeleteError,
    } = this.props;

    const {
      eventAttendees,
      attendances
    } = this.state;

    const reserves = this.getReserves(eventAttendees);
    const attendees = this.getAttendees(eventAttendees);

    return (
      <div className={'event'} data-event-id={id}>
        <h4>{name}</h4>
        <p>{description }</p>
        <p><b>{`Start: ${startsAt.split('T')[0]}`}</b></p>
        {attendees && attendees.length > 0 && (
          <Fragment>
            <h5>Attendees</h5>
            <EventListAttendees attendees={attendees} isReserves={false} eventId={id} />
          </Fragment>
        )}
        {reserves && reserves.length > 0 && (
          <Fragment>
            <h5 className="spacer-15">Reserves</h5>
            <EventListAttendees attendees={reserves} isReserves={true} eventId={id} />
          </Fragment>
        )}
        <EventActions
          adminSection={adminSection}
          onDeleteSuccess={onDeleteSuccess}
          onDeleteError={onDeleteError}
          onUpdateAttendance={this.handleUpdateAttendance}
          onCancelError={this.handleCancelError}
          onJoinError={this.handleJoinError}
          attendances={attendances}
          id={id}
        />
      </div>
    );
  }
}

export default Event;
