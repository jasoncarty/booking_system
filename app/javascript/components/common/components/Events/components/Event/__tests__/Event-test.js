import React from 'react';
import { create } from 'react-test-renderer';

import Event from './../index';

jest
  .mock('./../../EventListAttendees', () => 'EventListAttendees')
  .mock('./../../EventActions', () => 'EventActions');

const defaultProps = {
  id: 1,
  name: 'Fake name',
  description: 'Fake description',
  startsAt: '098739084Tlkjhasdf',
  eventAttendees: [],
  attendances: [1,2,3],
  onDeleteError: () => {},
  onDeleteSuccess: () => {},
  adminSection: true
};

describe('<Event />', () => {
  it('matches snapshot if there no attendees', () => {
    const element = create(<Event {...defaultProps} />);
    expect(element).toMatchSnapshot();
  });

  it('matches snapshot if there are attendees but no reserves', () => {
    const props = {
      ...defaultProps,
      eventAttendees: [
        { reserve: false },
        { reserve: false },
        { reserve: false }
      ]
    };
    const element = create(<Event {...props} />);
    expect(element).toMatchSnapshot();
  });

  it('matches snapshot if there are attendees and reserves', () => {
    const props = {
      ...defaultProps,
      eventAttendees: [
        { reserve: false },
        { reserve: true },
        { reserve: false }
      ]
    };
    const element = create(<Event {...props} />);
    expect(element).toMatchSnapshot();
  });

  it('tests exclusion/inclusion from event', () => {
    const originalAttendees = [
      { reserve: false },
      { reserve: true },
      { reserve: false }
    ];

    const props = {
      ...defaultProps,
      eventAttendees: originalAttendees
    };

    const element = create(<Event {...props} />);
    const instance = element.getInstance();
    expect(instance.state.eventAttendees).toEqual(originalAttendees);

    const { attendances } = defaultProps;
    const newAttendees = [
      { reserve: true },
      { reserve: false }
    ];

    instance.handleUpdateAttendance({
      data: {
        eventAttendees: newAttendees,
        attendances
      }
    });
    expect(instance.state.eventAttendees).toEqual(newAttendees);
  });
});
