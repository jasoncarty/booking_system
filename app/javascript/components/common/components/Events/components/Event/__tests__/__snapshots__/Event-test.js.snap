// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Event /> matches snapshot if there are attendees and reserves 1`] = `
<div
  className="event"
  data-event-id={1}
>
  <h4>
    Fake name
  </h4>
  <p>
    Fake description
  </p>
  <p>
    <b>
      Start: 098739084
    </b>
  </p>
  <h5>
    Attendees
  </h5>
  <EventListAttendees
    attendees={
      Array [
        Object {
          "reserve": false,
        },
        Object {
          "reserve": false,
        },
      ]
    }
    eventId={1}
    isReserves={false}
  />
  <h5
    className="spacer-15"
  >
    Reserves
  </h5>
  <EventListAttendees
    attendees={
      Array [
        Object {
          "reserve": true,
        },
      ]
    }
    eventId={1}
    isReserves={true}
  />
  <EventActions
    adminSection={true}
    attendances={
      Array [
        1,
        2,
        3,
      ]
    }
    id={1}
    onCancelError={[Function]}
    onDeleteError={[Function]}
    onDeleteSuccess={[Function]}
    onJoinError={[Function]}
    onUpdateAttendance={[Function]}
  />
</div>
`;

exports[`<Event /> matches snapshot if there are attendees but no reserves 1`] = `
<div
  className="event"
  data-event-id={1}
>
  <h4>
    Fake name
  </h4>
  <p>
    Fake description
  </p>
  <p>
    <b>
      Start: 098739084
    </b>
  </p>
  <h5>
    Attendees
  </h5>
  <EventListAttendees
    attendees={
      Array [
        Object {
          "reserve": false,
        },
        Object {
          "reserve": false,
        },
        Object {
          "reserve": false,
        },
      ]
    }
    eventId={1}
    isReserves={false}
  />
  <EventActions
    adminSection={true}
    attendances={
      Array [
        1,
        2,
        3,
      ]
    }
    id={1}
    onCancelError={[Function]}
    onDeleteError={[Function]}
    onDeleteSuccess={[Function]}
    onJoinError={[Function]}
    onUpdateAttendance={[Function]}
  />
</div>
`;

exports[`<Event /> matches snapshot if there no attendees 1`] = `
<div
  className="event"
  data-event-id={1}
>
  <h4>
    Fake name
  </h4>
  <p>
    Fake description
  </p>
  <p>
    <b>
      Start: 098739084
    </b>
  </p>
  <EventActions
    adminSection={true}
    attendances={
      Array [
        1,
        2,
        3,
      ]
    }
    id={1}
    onCancelError={[Function]}
    onDeleteError={[Function]}
    onDeleteSuccess={[Function]}
    onJoinError={[Function]}
    onUpdateAttendance={[Function]}
  />
</div>
`;
